File: models/reddit.py
"""Models for the Reddit scraper service.

The service wraps an Apify Reddit-scraper Actor behind a small first-party
surface: callers pass a search term (or explicit Reddit URLs) and get back the
matched items. Items are heterogeneous (posts, comments, communities, users
depending on the query) and each Actor names its fields differently, so the
service maps every item onto a stable, actor-agnostic ``NormalizedRedditItem``
(snake_case, engagement fields nullable) and keeps the untouched Actor item
under ``raw`` so nothing is lost. See ``services/reddit_svc.py`` for the
per-actor mapping layer that populates this shape.
"""

from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

RedditSort = Literal["relevance", "hot", "top", "new", "rising", "comments"]
RedditTime = Literal["all", "hour", "day", "week", "month", "year"]
RedditItemType = Literal["post", "comment", "community", "user"]


class RedditScrapeInput(BaseModel):
    """Query for the Reddit scraper.

    Provide ``search`` (optionally narrowed to a ``subreddit``) or one or more
    ``start_urls``. At least one of the two is required.
    """

    search: str | None = Field(
        default=None,
        description="Search term to look up on Reddit. Omit when using start_urls.",
    )
    subreddit: str | None = Field(
        default=None,
        description="Restrict a search to one community, e.g. 'programming'. "
        "Ignored when search is not set.",
    )
    start_urls: list[str] = Field(
        default_factory=list,
        description="Explicit Reddit post/community/user URLs to scrape directly.",
    )
    sort: RedditSort = Field(default="new", description="Ordering applied to a search.")
    time_filter: RedditTime | None = Field(
        default=None, description="Restrict posts to a recency window (posts only)."
    )
    max_items: int = Field(
        default=10,
        ge=1,
        le=1000,
        description="Maximum number of dataset items to return.",
    )
    include_comments: bool = Field(
        default=False, description="Also scrape comments on matched posts."
    )
    include_nsfw: bool = Field(default=False, description="Include NSFW results.")
    include_media_links: bool = Field(
        default=False,
        description="Extract engagement fields (upvotes, comment count, upvote "
        "ratio) and media URLs. Off by default: the Actor's fast RSS mode omits "
        "these, so turning this on switches to a slower detailed scrape. Enable "
        "it when downstream ranking needs reach/engagement signal.",
    )

    @model_validator(mode="after")
    def _require_a_target(self) -> "RedditScrapeInput":
        # Normalize a blank/whitespace-only search to None so it fails the
        # check below instead of being sent to Apify as a useless query.
        if self.search is not None:
            self.search = self.search.strip() or None
        if not self.search and not self.start_urls:
            raise ValueError(
                "Provide either 'search' or at least one 'start_urls' entry."
            )
        return self


class NormalizedRedditItem(BaseModel):
    """A Reddit post/comment/community/user in a stable, actor-agnostic shape.

    Every field is nullable: a given Actor (or item kind) may omit any of them,
    and an absent engagement count stays ``None`` rather than being faked as
    ``0`` so "unknown" stays distinct from "zero". The untouched Actor item is
    preserved under ``raw``.
    """

    id: str | None = Field(default=None, description="Actor item id, if any.")
    type: RedditItemType | None = Field(
        default=None, description="Item kind: post | comment | community | user."
    )
    title: str | None = Field(default=None, description="Post title.")
    body: str | None = Field(default=None, description="Post selftext or comment body.")
    author: str | None = Field(default=None, description="Author username.")
    subreddit: str | None = Field(
        default=None, description="Community name (without the 'r/' prefix if given)."
    )
    url: str | None = Field(default=None, description="Canonical URL for the item.")
    permalink: str | None = Field(
        default=None, description="Reddit permalink path, when derivable."
    )
    created_at: str | None = Field(
        default=None, description="Creation time as an ISO8601 string."
    )
    score: int | None = Field(
        default=None, description="Net upvotes; None when the Actor omits it."
    )
    num_comments: int | None = Field(
        default=None, description="Comment count; None when the Actor omits it."
    )
    upvote_ratio: float | None = Field(
        default=None, description="Upvote ratio 0..1; None when the Actor omits it."
    )
    over_18: bool | None = Field(
        default=None, description="NSFW flag; None when the Actor omits it."
    )
    num_crossposts: int | None = Field(
        default=None, description="Crosspost count; None when the Actor omits it."
    )
    raw: dict[str, Any] = Field(
        default_factory=dict, description="The untouched Actor dataset item."
    )


class RedditScrapeResult(BaseModel):
    """Normalized items returned by the Actor run.

    ``items`` are mapped onto the stable ``NormalizedRedditItem`` shape; each
    carries the original Actor item under ``raw`` so no upstream data is lost.
    """

    count: int = Field(description="Number of items returned.")
    items: list[NormalizedRedditItem] = Field(
        default_factory=list,
        description="Normalized dataset items (raw Actor item preserved per item).",
    )


class RedditScrapeStartResult(BaseModel):
    """Handle for an asynchronous Actor run started by ``reddit_scrape_start``.

    The run keeps executing on Apify after this returns; poll
    ``reddit_scrape_fetch`` with ``run_id`` until its status is terminal. Use
    the async path for slow queries (keyword search enumeration) that would
    exceed a synchronous call's client timeout; ``reddit_scrape`` stays the
    fast path for listing/comment pulls.
    """

    run_id: str = Field(description="Apify actor-run id; pass to reddit_scrape_fetch.")
    dataset_id: str = Field(
        description="Default dataset id for the run (items land here when it finishes)."
    )
    status: str = Field(
        description="Initial run status (e.g. READY or RUNNING) - not yet terminal."
    )


class RedditScrapeFetchInput(BaseModel):
    """Poll input for ``reddit_scrape_fetch``: the run id from a prior start."""

    run_id: str = Field(
        min_length=1,
        pattern=r"^[A-Za-z0-9_-]+$",
        description="Apify actor-run id returned by reddit_scrape_start.",
    )


class RedditScrapeFetchResult(BaseModel):
    """Current state of an async run, with items once it has SUCCEEDED.

    While the run is non-terminal (READY/RUNNING/*ING) ``items`` is empty and
    the caller should poll again. On SUCCEEDED, ``items`` holds the normalized
    dataset. On a terminal failure (FAILED/TIMED-OUT/ABORTED) the failing
    ``status`` is returned with empty ``items`` so the caller stops polling.
    """

    status: str = Field(description="Apify run status at poll time.")
    count: int = Field(description="Number of items returned (0 until SUCCEEDED).")
    items: list[NormalizedRedditItem] = Field(
        default_factory=list,
        description="Normalized dataset items once SUCCEEDED (raw item preserved).",
    )

File: servers/reddit/src/reddit.ts
/**
 * Pure, runtime-agnostic helpers for the reddit MCP server.
 *
 * Nothing here touches Cloudflare bindings, the MCP SDK, or `fetch` on purpose:
 * the real decisions (what target is valid, how the first-party input maps onto
 * the Apify Actor's schema, what a well-formed dataset looks like, and how each
 * Actor's item is normalized onto the stable NormalizedRedditItem shape) are the
 * part worth unit-testing, and keeping them dependency-free lets `bun test`
 * exercise them offline with no workerd / network.
 *
 * Mirrors the Python service `services/reddit_svc.py` at the repo root, which
 * wraps the same Actor for the CLI/HTTP/stdio transports.
 */

// Per-Actor field maps live in ./adapters; normalizeItem resolves one through
// this helper. The import is call-time only (used inside normalizeItem), so the
// reddit <-> adapters cycle is safe: neither module touches the other at init.
import { fieldMapForActor } from "./adapters";

export const APIFY_BASE = "https://api.apify.com/v2";

/**
 * The reddit-scraper Actor the fleet server runs by default. Every supported
 * Actor maps onto the same normalized item shape via a per-Actor adapter (input
 * builder + output field map; see below), so the choice is a cost/reliability
 * decision, not a contract change:
 *
 *  - fatihtahta/reddit-scraper-search-fast ("Enterprise Grade") - the default.
 *    In production it ran ~2.5x cheaper per run than trudax/reddit-scraper-lite
 *    ($0.042 vs $0.105) and, unlike lite, never TIMED-OUT on keyword searches
 *    (lite timed out on ~16% of runs, each still billed). Reddit-native
 *    snake_case output.
 *  - trudax/reddit-scraper-lite - the prior default, kept fully supported as an
 *    instant rollback: set APIFY_ACTOR_ID to it (no redeploy of logic needed).
 *
 * Tilde form is the URL-safe "username~name".
 */
export const DEFAULT_ACTOR_ID = "fatihtahta~reddit-scraper-search-fast";

/**
 * Cap the synchronous Apify run. Lower than the Python service's 300s: a Worker
 * subrequest has a tighter wall-clock budget than a long-lived Python process,
 * and small `max_items` queries return well within this. Bump `max_items` down,
 * not this up, for heavier scrapes.
 */
export const RUN_TIMEOUT_S = 120;

export type RedditSort = "relevance" | "hot" | "top" | "new" | "rising" | "comments";
export type RedditTime = "all" | "hour" | "day" | "week" | "month" | "year";

export interface RedditScrapeArgs {
  search?: string;
  subreddit?: string;
  start_urls?: string[];
  sort?: RedditSort;
  time_filter?: RedditTime;
  max_items?: number;
  include_comments?: boolean;
  include_nsfw?: boolean;
  include_media_links?: boolean;
}

/**
 * Normalize a search term: trim, and treat blank/whitespace-only as absent so a
 * useless `searches: ["  "]` is never sent to Apify (it must instead hit the
 * "provide a search or URLs" error).
 */
export function normalizeSearch(search: string | undefined): string | undefined {
  if (search === undefined) return undefined;
  const trimmed = search.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

/** A query needs at least one target: a real search term or one start URL. */
export function hasTarget(args: RedditScrapeArgs): boolean {
  return Boolean(normalizeSearch(args.search)) || (args.start_urls?.length ?? 0) > 0;
}

// Per-Actor input building (and output field maps) live in ./adapters, keyed by
// Actor slug. buildActorInput / fieldMapForActor there resolve from one registry
// so the input we send and the output we normalize can never disagree.

/** Strip any trailing slashes from the API base so path joins never double up. */
function normalizeBase(base: string): string {
  return base.replace(/\/+$/, "");
}

/** The synchronous run-and-fetch-dataset endpoint for an Actor. */
export function runSyncUrl(actorId: string, base: string = APIFY_BASE): string {
  return `${normalizeBase(base)}/acts/${actorId}/run-sync-get-dataset-items`;
}

// --- Async run + poll -------------------------------------------------------
//
// runSyncUrl blocks one subrequest on the whole run, which the innermost MCP
// client can cut off long before a slow keyword search finishes. The endpoints
// and parsers below decouple start from poll: `reddit_scrape_start` POSTs to
// runsUrl for a run id, `reddit_scrape_fetch` GETs runStatusUrl and, once the
// run has SUCCEEDED, datasetItemsUrl. Mirrors services/reddit_svc.py.

/** Non-blocking run-enqueue endpoint: POST returns a run without waiting on it. */
export function runsUrl(actorId: string, base: string = APIFY_BASE): string {
  return `${normalizeBase(base)}/acts/${actorId}/runs`;
}

/**
 * Run-status endpoint: GET returns the run's `{ status, defaultDatasetId }`.
 * `runId` is caller-supplied, so encode it as a single path segment - a value
 * carrying `/` or `?` can never escape into the path or query.
 */
export function runStatusUrl(runId: string, base: string = APIFY_BASE): string {
  return `${normalizeBase(base)}/actor-runs/${encodeURIComponent(runId)}`;
}

/** Dataset items endpoint for a finished run; `datasetId` is encoded per segment. */
export function datasetItemsUrl(datasetId: string, base: string = APIFY_BASE): string {
  return `${normalizeBase(base)}/datasets/${encodeURIComponent(datasetId)}/items`;
}

/**
 * Apify run statuses. SUCCEEDED is the only terminal-success state; the failures
 * are terminal but yield no items; everything else means "still running, poll
 * again". https://docs.apify.com/platform/actors/running/runs-and-builds#lifecycle
 */
export const SUCCEEDED = "SUCCEEDED";
export const TERMINAL_FAILURE: ReadonlySet<string> = new Set(["FAILED", "TIMED-OUT", "ABORTED"]);

export type RunStartResult =
  | { ok: true; run_id: string; dataset_id: string; status: string }
  | { ok: false; error: string };

/** Parse an Apify run-enqueue envelope `{ data: { id, defaultDatasetId, status } }`. */
export function parseRunStart(json: unknown): RunStartResult {
  const data = extractData(json);
  if (!data) return { ok: false, error: `unexpected Apify response shape: ${typeof json}` };
  const run_id = asStr(data.id);
  const dataset_id = asStr(data.defaultDatasetId);
  const status = asStr(data.status);
  if (!run_id || !dataset_id || !status) {
    return { ok: false, error: "Apify run response is missing id/defaultDatasetId/status" };
  }
  return { ok: true, run_id, dataset_id, status };
}

export type RunStatusResult =
  | { ok: true; status: string; dataset_id: string | null }
  | { ok: false; error: string };

/** Parse an Apify run-status envelope `{ data: { status, defaultDatasetId } }`. */
export function parseRunStatus(json: unknown): RunStatusResult {
  const data = extractData(json);
  if (!data) return { ok: false, error: `unexpected Apify response shape: ${typeof json}` };
  const status = asStr(data.status);
  if (!status) return { ok: false, error: "Apify run response is missing status" };
  return { ok: true, status, dataset_id: asStr(data.defaultDatasetId) };
}

function extractData(json: unknown): Record<string, unknown> | null {
  if (json === null || typeof json !== "object" || Array.isArray(json)) return null;
  const data = (json as Record<string, unknown>).data;
  if (data === null || typeof data !== "object" || Array.isArray(data)) return null;
  return data as Record<string, unknown>;
}

export type DatasetResult =
  | { ok: true; items: Record<string, unknown>[] }
  | { ok: false; error: string };

/**
 * Validate a parsed Apify response: it must be a JSON array whose every element
 * is a plain object. Anything else is surfaced as an error rather than handed
 * back as malformed data.
 */
export function validateDatasetItems(json: unknown): DatasetResult {
  if (!Array.isArray(json)) {
    return { ok: false, error: `unexpected Apify response shape: ${typeof json}` };
  }
  for (const item of json) {
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      return { ok: false, error: "Apify returned a dataset item that is not an object" };
    }
  }
  return { ok: true, items: json as Record<string, unknown>[] };
}

// --- Normalization ---------------------------------------------------------
//
// Callers never depend on a specific Actor's raw keys: normalizeItem maps every
// item onto the stable NormalizedRedditItem shape using that Actor's field map
// (a normalized field -> ordered candidate source keys; first present wins),
// resolved via fieldMapForActor in ./adapters. Mirrors the Python normalizer in
// services/reddit_normalize.py.

export type RedditItemType = "post" | "comment" | "community" | "user";

export interface NormalizedRedditItem {
  id: string | null;
  type: RedditItemType | null;
  title: string | null;
  body: string | null;
  author: string | null;
  subreddit: string | null;
  url: string | null;
  permalink: string | null;
  created_at: string | null;
  score: number | null;
  num_comments: number | null;
  upvote_ratio: number | null;
  over_18: boolean | null;
  num_crossposts: number | null;
  raw: Record<string, unknown>;
}

/**
 * A field map keys a normalized field to an ordered list of candidate source
 * keys. Defined here (beside NormalizedRedditItem) but populated per-Actor in
 * ./adapters.
 */
export type FieldMap = Record<keyof Omit<NormalizedRedditItem, "raw">, string[]>;

/** Raw type/kind discriminators (incl. Reddit's t1/t3/t5/t2 codes) -> our literal. */
const TYPE_ALIASES: Record<string, RedditItemType> = {
  post: "post",
  link: "post",
  t3: "post",
  comment: "comment",
  t1: "comment",
  community: "community",
  subreddit: "community",
  sr: "community",
  t5: "community",
  user: "user",
  account: "user",
  t2: "user",
};

function firstPresent(raw: Record<string, unknown>, keys: string[]): unknown {
  for (const key of keys) {
    const value = raw[key];
    if (value !== undefined && value !== null) return value;
  }
  return null;
}

function asStr(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  return typeof value === "string" ? value : String(value);
}

function asNumber(value: unknown): number | null {
  // Reject booleans: `typeof true === "boolean"`, but a flag is never a count.
  if (typeof value === "number") return Number.isFinite(value) ? value : null;
  if (typeof value === "string") {
    const parsed = Number(value.trim());
    return value.trim() !== "" && Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

function asInt(value: unknown): number | null {
  const n = asNumber(value);
  return n === null ? null : Math.trunc(n);
}

function asBool(value: unknown): boolean | null {
  if (typeof value === "boolean") return value;
  if (typeof value === "number") return value !== 0;
  if (typeof value === "string") {
    const low = value.trim().toLowerCase();
    if (["true", "1", "yes"].includes(low)) return true;
    if (["false", "0", "no", ""].includes(low)) return false;
  }
  return null;
}

function asIso(value: unknown): string | null {
  // A Unix epoch (seconds) becomes an ISO8601 UTC string; a string passes
  // through unchanged (trudax already emits ISO8601).
  if (typeof value === "number" && Number.isFinite(value)) {
    return new Date(value * 1000).toISOString();
  }
  if (typeof value === "string") return value.trim() || null;
  return null;
}

function asType(value: unknown): RedditItemType | null {
  if (typeof value !== "string") return null;
  return TYPE_ALIASES[value.trim().toLowerCase()] ?? null;
}

function cleanSubreddit(value: unknown): string | null {
  const text = asStr(value)?.trim();
  if (!text) return null;
  for (const prefix of ["/r/", "r/"]) {
    if (text.toLowerCase().startsWith(prefix)) return text.slice(prefix.length) || null;
  }
  return text;
}

function derivePermalink(explicit: unknown, url: unknown): string | null {
  // Prefer an explicit permalink; otherwise recover the path from a reddit.com
  // URL so downstream consumers get a stable permalink even when the Actor only
  // returns a full URL.
  const text = asStr(explicit);
  if (text) return text;
  const full = asStr(url);
  if (!full) return null;
  let parsed: URL;
  try {
    parsed = new URL(full);
  } catch {
    return null;
  }
  const host = parsed.hostname.toLowerCase();
  if (parsed.pathname && (host === "reddit.com" || host.endsWith(".reddit.com"))) {
    return parsed.pathname;
  }
  return null;
}

/** Map one raw Actor item onto the stable NormalizedRedditItem shape. */
export function normalizeItem(raw: Record<string, unknown>, actorId: string): NormalizedRedditItem {
  const fmap = fieldMapForActor(actorId);
  const pick = (field: keyof FieldMap): unknown => firstPresent(raw, fmap[field]);
  return {
    id: asStr(pick("id")),
    type: asType(pick("type")),
    title: asStr(pick("title")),
    body: asStr(pick("body")),
    author: asStr(pick("author")),
    subreddit: cleanSubreddit(pick("subreddit")),
    url: asStr(pick("url")),
    permalink: derivePermalink(pick("permalink"), pick("url")),
    created_at: asIso(pick("created_at")),
    score: asInt(pick("score")),
    num_comments: asInt(pick("num_comments")),
    upvote_ratio: asNumber(pick("upvote_ratio")),
    over_18: asBool(pick("over_18")),
    num_crossposts: asInt(pick("num_crossposts")),
    raw,
  };
}

export function normalizeItems(
  items: Record<string, unknown>[],
  actorId: string,
): NormalizedRedditItem[] {
  return items.map((item) => normalizeItem(item, actorId));
}
